### PR TITLE
ARM support and tools_def.txt cleanup.

### PR DESCRIPTION
**Pull request by MikhailKrichanov**

### CODE REVIEW

File: ArmVirtPkg/ArmVirtKvmTool.dsc [hunk 1]
```diff
@@ -94,11 +94,6 @@
 
 [BuildOptions]
   *_*_*_CC_FLAGS = -D DISABLE_NEW_DEPRECATED_INTERFACES
-  #
-  # We need to avoid jump tables in SEC and BASE modules, so that the PE/COFF
-  # self-relocation code itself is guaranteed to be position independent.
-  #
-  GCC:*_*_*_CC_XIPFLAGS = -fno-jump-tables
```
Comment (mhaeuser) [inline ArmVirtPkg/ArmVirtKvmTool.dsc:101]: Why can we just drop this?

File: PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf [hunk 2]
```diff
@@ -39,7 +39,5 @@
 
 [BuildOptions.common]
   MSFT:*_*_*_DLINK_FLAGS  = /DLL /SUBSYSTEM:CONSOLE /VERSION:1.0
-  MSFT:*_*_*_GENFW_FLAGS = --keepoptionalheader
 
-  GCC:*_*_AARCH64_GENFW_FLAGS = --prm
```
Comment (mhaeuser) [inline PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf:44]: Why are we removing the PRM flag from PRM modules? All this does is make life harder if we at some point want to support PRM, because we have to go back through such commits to see which important flags were deleted. This also makes clear to me that GENFW_FLAGS should not be removed, as there is good reason for module definitions to modify it.

File: ArmVirtPkg/ArmVirtKvmTool.dsc [hunk 3]
```diff
@@ -352,17 +347,14 @@
   ArmPkg/Drivers/ArmPciCpuIo2Dxe/ArmPciCpuIo2Dxe.inf {
     <LibraryClasses>
       NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
-      NULL|OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf
```
Comment (mhaeuser) [inline ArmVirtPkg/ArmVirtKvmTool.dsc:355]: Why is this removed?

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 1 above)
Comment (MikhailKrichanov) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:101]: [-fno-jump-tables](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html)
Do not use jump tables for switch statements even where it would be more efficient than other code generation strategies. This option is of use in conjunction with -fpic or -fPIC for building code that forms part of a dynamic linker and cannot reference the address of a jump table. On some targets, jump tables do not require a GOT and this option is not needed. 

All images generated by ImageTool are position dependent [(-fno-pic -fno-pie)](https://github.com/acidanthera/audk/blob/master/BaseTools/Conf/tools_def.template#L1880) 

I [removed](https://github.com/acidanthera/audk/pull/27/commits/cc272bb838b65735b173bb54581245af2630f415) the use case of that flag entirely and am still  able to boot Debian 11 for aarch64/arm.

File: PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf (hunk 2 above)
Comment (MikhailKrichanov) [reply PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf:44]: We are removing GENFW_FLAGS because we have already removed GenFw from everywhere except BaseTools directory itself. If one day you decide to broaden ImageTool functionality by implementing some of the GenFw features, your life will be as hard as typing grep command in the console over the original edk2 repository (in order to find out which important flags you need).

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 1 above)
Comment (mhaeuser) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:101]: Position-dependent codegen actually makes the problem worse, as RIP-based addressing is less likely to be used. The description of the flag is ambiguous - it does not explicitly state the option is without effect for -fno-pic, only that it makes sense to be used with -fPIC to avoid absolute addressing (that's why the last sentence mentions GOT). While the binaries as a whole may have relocs for absolute addresses, the idea (and requirement!) of self-relocation is that the early code performing the relocation is in fact position-independent without relocs.

Something that relies on this and in fact is not -fno-pic -fno-pie is StandaloneMmCore [1]. StandaloneMmCoreEntryPointLib self-relocates the image very soon after start [2]. I'm not sure whether it works as-is as I never got to test ARM targets before. Booting one OS on one target without verifying all platforms with self-relocating modules is not a sufficient test at all to remove such functionality.

StandaloneMmCore is not affected by this change, as it was specific to SEC and PEI modules, but I do not have a good overview over all ARM platform trees to tell whether really all of them are safe to build now with this removed. StandaloneMmCore just is the one example I'm somewhat familiar with how it works.

[1] https://github.com/acidanthera/audk/blob/b877d75f50b3a01b09f62ee3265adfa5c6da9622/StandaloneMmPkg/Core/StandaloneMmCore.inf#L80-L87

[2] https://github.com/acidanthera/audk/blob/b877d75f50b3a01b09f62ee3265adfa5c6da9622/StandaloneMmPkg/Library/StandaloneMmCoreEntryPoint/Arm/StandaloneMmCoreEntryPoint.c#L351-L400

File: PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf (hunk 2 above)
Comment (mhaeuser) [reply PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf:44]: Yes, because I am aware of this particular change. There have been similar changes in the past I was not aware of and I wasted significant time trying to understand the status quo in AUDK, before deciding to check edk2 and finding the missing pieces.

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 3 above)
Comment (MikhailKrichanov) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:355]: To fix CLANGDWARF build for ArmVirtKvmTool.dsc. 

ld.lld: error: duplicate symbol: PciExpressRegisterForRuntimeAccess
>>> defined in /home/mikhail/Amaranth/UDK/Build/ArmVirtKvmTool-AARCH64/DEBUG_CLANGDWARF/AARCH64/MdePkg/Library/BasePciExpressLib/BasePciExpressLib/OUTPUT/BasePciExpressLib.lib(PciExpressLib.obj)
>>> defined in /home/mikhail/Amaranth/UDK/Build/ArmVirtKvmTool-AARCH64/DEBUG_CLANGDWARF/AARCH64/OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib/OUTPUT/BaseCachingPciExpressLib.lib(PciExpressLib.obj)

...

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 1 above)
Comment (vit9696) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:101]: This is an interesting and fairly dangerous example. The comment in the .dsc is very misleading.

@mhaeuser, does it really mean that SEC/PEI code on ARM may be moved without updating the relocations? To me this violates the mere basics of image loading…


If there is such a requirement, I request:

- The flag is to be kept.
- The comment changed to explain that no relocations are allowed in SEC/PEI on ARM due to moving code. @mhaeuser, I would welcome if you suggest more clear wording.
- No relocations must be enforced by ImageTool. I am not exactly sure how to do this, but this should be done.

File: PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf (hunk 2 above)
Comment (vit9696) [reply PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf:44]: To be honest, PRMs are disgusting, and I would easily suggest removing them all along with the whole Pkg.

However, I agree with @mhaeuser that we should not make modifications to packages we neither test, nor build. PrmPkg is dead for us, and we do not test it. Please avoid making changes in it.

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 1 above)
Comment (mhaeuser) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:101]: @vit9696 No, I think there's a slight misunderstanding. No relocations are allowed *in the self-relocation code*. As you're aware, I'm not terribly familiar with ARM, but I was told by Ard or Leif (don't remember) that for StandaloneMmCore and other "initial" modules (which SEC would be, thus the flag being SEC/PEI only makes sense to me), they may indeed have load addresses unknown at load-time (which is not a concern with x86 SEC, obviously) - whether that depends on the exact hardware platform, I don't know. The idea, as you can see in [2], is that the module relocates itself. This process, from the entry point till `UefiImageRelocateImageInplaceForExecution()` returns, must work position-independently without relocs. As after it returned relocs have been processed, code that relies on relocs is safe from that point onwards.

As you can see in [1], it passes `-Wl,-z,text` to the linker, so it will warn about any relocs to .text and other read-only segments. There may of course in theory be some dependency on .data (et al.) relocs, but our loader does not use global variables, and I suppose that is the "convention" to make this safe. I don't think ImageTool needs to be involved with how things work right now, at least for StandaloneMmCore.

File: PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf (hunk 2 above)
Comment (mhaeuser) [reply PrmPkg/Samples/PrmSampleAcpiParameterBufferModule/PrmSampleAcpiParameterBufferModule.inf:44]: @vit9696 Yes, removing the entire thing is fine to me, as then it's clear the tree has nothing interesting for it and I need to look in tianocore/edk2. But if the package remains and only the important build flags are gone, this is not clear at all (without prior knowledge of this commit), and I would waste time trying to understand the AUDK snapshot, before resorting to tianocore/edk2 only eventually. It's partial removals that are a pain.

File: ArmVirtPkg/ArmVirtKvmTool.dsc (hunk 3 above)
Comment (MikhailKrichanov) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:355]: `PciExpressLib` is defined 2 times: 
1.  In `ArmVirtPkg/ArmVirt.dsc.inc` as  `PciExpressLib|OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf`
2.  In `ArmVirtPkg/ArmVirtKvmTool.dsc` as `PciExpressLib|MdePkg/Library/BasePciExpressLib/BasePciExpressLib.inf`

Removing definition in `ArmVirtPkg/ArmVirtKvmTool.dsc` leads to:
```
/home/mikhail/Amaranth/UDK/ArmVirtPkg/ArmVirtKvmTool.dsc(...): error 1001: Module type [UEFI_APPLICATION] is not supported by library instance [/home/mikhail/Amaranth/UDK/OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf]
	consumed by library instance [/home/mikhail/Amaranth/UDK/OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf] which is consumed by module [/home/mikhail/Amaranth/UDK/ShellPkg/Application/Shell/Shell.inf]
```

Comment (mhaeuser) [reply ArmVirtPkg/ArmVirtKvmTool.dsc:355]: Well, that’s not how DSC LibraryClasses work. Definitions of the same class *never* duplicate, they override each other by some odd rules (generally, the more specific scope takes precedence for when it applies).

You get the duplication because the override you removed is NULL, thus not tied to any specific LibraryClass and is always linked. Most likely those modules don’t have explicit dependencies on PciExpressLib, but some change in AUDK caused an implicit dependency, thus the default implementation clashes with this NULL one. That should probably investigated first, whether the new implicit dependency is legitimate.

if it is, it should be sufficient to change the LibraryClass name from NULL to PciExpressLib here. I don’t see a reason why you would force-link this library, as the constructor has no interesting side-effects.

File: BaseTools/Conf/tools_def.template [hunk 4]
```diff
@@ -1874,25 +719,19 @@ NOOPT_*_*_OBJCOPY_ADDDEBUGFLAG     = --add-gnu-debuglink=$(DEBUG_DIR)/$(MODULE_N
 DEFINE GCC_ALL_CC_FLAGS            = -g -Os -fshort-wchar -fno-builtin -fno-strict-aliasing -Wall -Werror -Wno-array-bounds -include AutoGen.h -fno-common
 DEFINE GCC_IA32_CC_FLAGS           = DEF(GCC_ALL_CC_FLAGS) -m32 -malign-double -freorder-blocks -freorder-blocks-and-partition -O2 -mno-stack-arg-probe
 DEFINE GCC_X64_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -mno-red-zone -Wno-address -mno-stack-arg-probe
-DEFINE GCC_ARM_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -march=armv7-a -mthumb -mlittle-endian -mabi=aapcs -fno-short-enums -funsigned-char -ffunction-sections -fdata-sections -fomit-frame-pointer -Wno-address -mthumb -mfloat-abi=soft -fno-pic -fno-pie
+DEFINE GCC_ARM_CC_FLAGS            = DEF(GCC_ALL_CC_FLAGS) -march=armv7-a -mthumb -mlittle-endian -mabi=aapcs -fno-short-enums -funsigned-char -ffunction-sections -fdata-sections -fomit-frame-pointer -Wno-address -mfloat-abi=soft -fno-pic -fno-pie
 DEFINE GCC_LOONGARCH64_CC_FLAGS    = DEF(GCC_ALL_CC_FLAGS) -mabi=lp64d -fno-asynchronous-unwind-tables -fno-plt -Wno-address -fno-short-enums -fsigned-char -ffunction-sections -fdata-sections
-DEFINE GCC_ARM_CC_XIPFLAGS         = -mno-unaligned-access
```
Comment (vit9696) [inline BaseTools/Conf/tools_def.template:1879]: Early stages on ARM/AArch64 execute with strict alignment enabled and will crash when compiler generates misaligned access. Please rename these to more accurate:

GCC_ARM_CC_SECPEI_FLAGS
GCC_AARCH_CC_SECPEI_FLAGS

And recover the build options:

ARM: -mno-unaligned-access
AARCH64: -mstrict-align -mgeneral-regs-only

File: ArmVirtPkg/ArmVirt.dsc.inc [hunk 5]
```diff
@@ -268,6 +261,12 @@
   # Disable deprecated APIs.
   #
   GCC:*_*_*_CC_FLAGS = -DDISABLE_NEW_DEPRECATED_INTERFACES
+
+!if $(PROTECT_RT_IMAGES_AARCH64_64K) == TRUE
+  [BuildOptions.common.EDKII.DXE_RUNTIME_DRIVER, BuildOptions.common.EDKII.SMM_CORE, BuildOptions.common.EDKII.DXE_SMM_DRIVER, BuildOptions.common.MM_CORE_STANDALONE, BuildOptions.common.MM_STANDALONE]
+    GCC:*_*_AARCH64_DLINK_FLAGS  = -z max-page-size=0x10000
+    MSFT:*_*_AARCH64_DLINK_FLAGS = /ALIGN:65536 /FILEALIGN:65536
+!endif
```
Comment (mhaeuser) [inline ArmVirtPkg/ArmVirt.dsc.inc:269]: Hmm, x86 defaults to "protection ready" values in tools_def without a DSC macro. @vit9696 Maybe the same should be done here?
